### PR TITLE
[SD sink] Various improvements to SD sink

### PR DESCRIPTION
**Pull request by piosz**
cherry-pick of #1752, #1755, #1767, #1757

### CODE REVIEW

File: metrics/sinks/stackdriver/stackdriver.go [hunk 1]
```diff
@@ -191,17 +219,114 @@ func (sink *StackdriverSink) ExportData(dataBatch *core.DataBatch) {
...
+
+	// Close the channel in order to cancel exporting routines.
+	close(requestQueue)
+
+	workersCompleted := 0
+	for {
+		select {
+		case <-completedQueue:
+			workersCompleted++
+			if workersCompleted == workers {
```
Comment (mwielgus) [inline metrics/sinks/stackdriver/stackdriver.go]: you can move this if condition to for, will be cleaner

File: metrics/sinks/stackdriver/stackdriver.go [hunk 2]
```diff
@@ -191,17 +219,114 @@ func (sink *StackdriverSink) ExportData(dataBatch *core.DataBatch) {
...
+	timeout := time.Duration(sink.batchExportTimeoutSec) * time.Second
+	timeoutSending := time.After(timeout)
+	timeoutCompleted := time.After(timeout)
+
+forloop:
+	for i, r := range requests {
+		select {
+		case requestQueue <- r:
+			// yet another request added to queue
+		case <-timeoutSending:
```
Comment (mwielgus) [inline metrics/sinks/stackdriver/stackdriver.go]: I would remove the timeout here and use buffered channels instead. There is no point in having timeout control in 2 places.

File: metrics/sinks/stackdriver/stackdriver.go (hunk 1 above)
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: will send a follow up PR - this is a pure cherry-pick

File: metrics/sinks/stackdriver/stackdriver.go (hunk 2 above)
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: Without this you won't be able to distinguish when a request was not picked at all vs. the case when it's in progress.

Comment (mwielgus) [inline metrics/sinks/stackdriver/stackdriver.go]: How useful this information is?

File: metrics/sinks/stackdriver/stackdriver.go (hunk 1 above)
Comment (mwielgus) [inline metrics/sinks/stackdriver/stackdriver.go]: ok.

File: metrics/sinks/stackdriver/stackdriver.go (hunk 2 above)
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: It's for accounting purposes: we want to track number of requests/timeseries successfully delivered to SD. Those which are not picked by any thread are clearly dropped.

Comment (mwielgus) [inline metrics/sinks/stackdriver/stackdriver.go]: Just FYI, switching to buffered channels will require you to change the stop method and introduce an extra check around line 295, so it is not a single line :/. 
Anyway as this is a cherry-pick I will not push for any changes but please consider refactoring it in the future.

**Pull request merged**